maxi-blocks/maxi-blocks
# Issue 4434 - Implement `handleSetAttributes` to IB

## Update 18/4/23
Some updates:
1. Fixed some issues related with the saving methodology with IB
2. Created unit tests

Some notes:
1. Need to clean the IB attributes when they are repeated; something similar to `styleProcessor`/`styleCleaner`

## Update 14/4/23
Some new updates and features:
1. Fixed the issue @Olekrut shared with responsive values
4. Created a system that updates IB styles from the affected block, which ensures the IB settings are always synchronized with the block to affect attributes.
5. Created kind of migrator on `maxiBlockComponent` that updates the old IB attributes.

# Description

This PR is a parallel and alternative version of #4708. As already commented to @FitseTLT, the #4708 PR is a correct and a really good implementation that fixes the issue, but in this concrete case uses a methodology that could be simpler having the correct context of the situation: #4434 issue is related with #3811, as the main problem is IB is saving sometimes more attributes than the ones it needs. If we watch the video @Olekrut shared on #4434, what we see is that we change a property not related with colour, then we change on the IB affected block the colour, but in IB, when triggering the action, the colour comes back to the old one. The reason is that we aren't saving just the modified attributes on IB in the `relation` object, we are saving all the attributes! So, following the requirements of #3811, this PR implements `handleSetAttributes` on IB to ensure we save the correct attributes. The list of implemented actions is:
1. Implements new filters on RelationControl that first cleans the attributes to save on `relation` object using `handleSetAttributes`, and to ensure a good result, ensures we save the related `unit` or `palette` attributes in case is necessary, and then gets the styles using a temporary object using the attributes defined on IB settings objects under the key `styleAttrs`.
6. As commented above, implemented a new entry on Relation settings object called `styleAttrs` that ensures that attributes will be used on the getting styles process for the IB cause normally this attributes are necessary inside the concrete helpers. These attributes aren't saved inside the IB object 👍 
7. Made a small esthetic change on the `src/extensions/styles/helpers/getAlignmentFlexStyles.js`, nothing big lol

PS: this implementation also works as an optimization, as it reduces the amount of attributes saved on `relations` object reducing the code-editor 👍 

Fixes #4434 and partially #3811 

# How Has This Been Tested?

To test this PR needs a bit of patience and love. The process is the same commented on #4434:
1. Create a Group Maxi and set inside a block to affect.
(*). Sometimes is necessary to set some attribute on the block to affect, like `box-shadow` options or button icon options
2. On Group Maxi set an interaction that contains colour, but don't change the colour, change other setting
3. Check frontend all work good.
8. On editor modify the main colour of the affected block (not from IB, from the affected block)
9. Check on frontend the colour is the selected one and the IB trigger doesn't change it
10. On Grup Maxi IB, change the colour of the block to affect
11. In frontend, with the IB action not triggered, the block should show the colour selected from the affected block, and on triggering the IB the colour should be modified to the one selected on the IB of the Group Maxi.

Here's a video:


https://user-images.githubusercontent.com/50477222/231695363-a83ee7c2-1562-43f9-99b1-1aed16a51f4c.mp4



# Test checklist



**_ Front/Back Testing _**

-   [ ] Test IB in general lol

**_ Pre-Code Testing _**

-   [ ] Test IB in general lol
